src/main/java/com/iota/iri/service/snapshot/impl/SnapshotProviderImpl.java

```diff
@@ -118,15 +119,38 @@ public Snapshot getLatestSnapshot() {
     }
 
     /**
-     * {@inheritDoc}
+     * {@inheritDoc}<br />
+     * <br />
+     * It first writes two temporary files and then renames them to avoid corruption of the written files in case of IRI
+     * crashes.<br />
      */
     @Override
     public void writeSnapshotToDisk(Snapshot snapshot, String basePath) throws SnapshotException {
         snapshot.lockRead();
 
+        File tmpStateFile = null;
+        File tmpMetaFile = null;
         try {
-            writeSnapshotStateToDisk(snapshot, basePath + ".snapshot.state");
-            writeSnapshotMetaDataToDisk(snapshot, basePath + ".snapshot.meta");
+            tmpStateFile = File.createTempFile("iri.", ".snapshot.state");
+            tmpMetaFile = File.createTempFile("iri.", ".snapshot.meta");
+
+            writeSnapshotStateToDisk(snapshot, tmpStateFile.getAbsolutePath());
+            writeSnapshotMetaDataToDisk(snapshot, tmpMetaFile.getAbsolutePath());
+
+            Files.move(Paths.get(tmpStateFile.getAbsolutePath()), Paths.get(basePath + ".snapshot.state"),
+                    StandardCopyOption.ATOMIC_MOVE);
```

Comment:
don't you need `REPLACE_EXISTING` as well?

Reply:
REPLACE_EXISTING is ignored if ATOMIC_MOVE is provided and it is plattform dependent if it replaces existing files (which is a bad thing i guess). 

Maybe the best way to solve this is to:

1. write new ".tmp" files in the same folder
2. rename old files to ".bkp"
3. rename ".tmp" files to their correct location
4. delete ".bkp" files

This way we would always have the backup files prior to the failed snapshot which could be renamed manually by the noder operator of this procedure ever fails.

Reply:
Maybe even always keep the bkp files instead of deleting them so node operators have a way to recover from really odd edge cases.

And just REPLACE_EXISTING the ".bkp" files while renaming them.